File: src/tools/ppt/spec.py
"""PptxGenJS renderer intermediate schema."""

from typing import Annotated, Literal, Union

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class SpecModel(BaseModel):
    model_config = ConfigDict(extra="forbid")

    @field_validator(
        "color", "fill", "line_color", "header_fill", "header_color",
        "border_color", "background", check_fields=False,
    )
    @classmethod
    def validate_hex_color(cls, value: str) -> str:
        normalized = value.removeprefix("#").upper()
        if len(normalized) != 6 or any(character not in "0123456789ABCDEF" for character in normalized):
            raise ValueError("color must be a 6-digit hexadecimal value")
        return normalized


class PositionedNode(SpecModel):
    x: float = Field(ge=0)
    y: float = Field(ge=0)
    w: float = Field(gt=0)
    h: float = Field(gt=0)


class TextNode(PositionedNode):
    type: Literal["text"] = "text"
    text: str
    font_size: float = Field(default=20, gt=0, le=200)
    font_face: str = "Microsoft YaHei"
    color: str = "222222"
    bold: bool = False
    align: Literal["left", "center", "right"] = "left"
    valign: Literal["top", "mid", "bottom"] = "top"
    margin: float = Field(default=0, ge=0)
    bullet: bool = False

    @field_validator("text")
    @classmethod
    def text_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("text must not be blank")
        return value


class ShapeNode(PositionedNode):
    type: Literal["shape"] = "shape"
    shape: Literal["rect", "roundRect", "line", "ellipse"] = "rect"
    fill: str = "FFFFFF"
    line_color: str = "FFFFFF"
    line_width: float = Field(default=0, ge=0)


class ImageNode(PositionedNode):
    type: Literal["image"] = "image"
    path: str
    transparency: int = Field(default=0, ge=0, le=100)

    @field_validator("path")
    @classmethod
    def path_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("path must not be blank")
        return value


class RasterLayerNode(ImageNode):
    type: Literal["raster"] = "raster"


class TableNode(PositionedNode):
    type: Literal["table"] = "table"
    rows: list[list[str]]
    font_size: float = Field(default=14, gt=0, le=100)
    color: str = "222222"
    header_fill: str = "1F4E78"
    header_color: str = "FFFFFF"
    border_color: str = "D9E2F3"

    @model_validator(mode="after")
    def validate_rows(self):
        if not self.rows or not self.rows[0]:
            raise ValueError("table rows must not be empty")
        width = len(self.rows[0])
        if any(len(row) != width for row in self.rows):
            raise ValueError("table rows must have equal column counts")
        return self


class ChartSeries(SpecModel):
    name: str
    values: list[float]


class ChartNode(PositionedNode):
    type: Literal["chart"] = "chart"
    chart_type: Literal["bar", "column", "line", "pie", "doughnut"] = "column"
    labels: list[str]
    series: list[ChartSeries]
    show_legend: bool = True
    show_title: bool = False
    title: str | None = None

    @model_validator(mode="after")
    def validate_chart_data(self):
        if not self.labels or not self.series:
            raise ValueError("chart labels and series must not be empty")
        if any(len(item.values) != len(self.labels) for item in self.series):
            raise ValueError("chart series values must match labels")
        return self


SlideNode = Annotated[
    Union[
        TextNode,
        ShapeNode,
        ImageNode,
        RasterLayerNode,
        TableNode,
        ChartNode,
    ],
    Field(discriminator="type"),
]


class SlideSpec(SpecModel):
    id: str
    layout: str | None = None
    background: str = "FFFFFF"
    nodes: list[SlideNode] = Field(default_factory=list)
    notes: str | None = None

    @field_validator("id")
    @classmethod
    def id_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("slide id must not be blank")
        return value


class SlideDeckSpec(SpecModel):
    version: Literal["1.0"] = "1.0"
    title: str
    width: float = Field(default=13.333, gt=0, le=100)
    height: float = Field(default=7.5, gt=0, le=100)
    author: str = "AID Work Agent"
    subject: str | None = None
    warnings: list[str] = Field(default_factory=list)
    slides: list[SlideSpec] = Field(min_length=1)

    @field_validator("title")
    @classmethod
    def title_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("deck title must not be blank")
        return value

    @model_validator(mode="after")
    def validate_node_bounds(self):
        for slide in self.slides:
            for node in slide.nodes:
                if node.x + node.w > self.width + 1e-6:
                    raise ValueError(f"node exceeds slide width on {slide.id}")
                if node.y + node.h > self.height + 1e-6:
                    raise ValueError(f"node exceeds slide height on {slide.id}")
        return self

File: src/tools/ppt/renderer-node/src/spec.ts
export type Align = "left" | "center" | "right";
export type VAlign = "top" | "mid" | "bottom";

export interface BaseNode {
  type: string;
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface TextNode extends BaseNode {
  type: "text";
  text: string;
  font_size: number;
  font_face: string;
  color: string;
  bold: boolean;
  align: Align;
  valign: VAlign;
  margin: number;
  bullet: boolean;
}

export interface ShapeNode extends BaseNode {
  type: "shape";
  shape: "rect" | "roundRect" | "line" | "ellipse";
  fill: string;
  line_color: string;
  line_width: number;
}

export interface ImageNode extends BaseNode {
  type: "image" | "raster";
  path: string;
  transparency: number;
}

export interface TableNode extends BaseNode {
  type: "table";
  rows: string[][];
  font_size: number;
  color: string;
  header_fill: string;
  header_color: string;
  border_color: string;
}

export interface ChartSeries {
  name: string;
  values: number[];
}

export interface ChartNode extends BaseNode {
  type: "chart";
  chart_type: "bar" | "column" | "line" | "pie" | "doughnut";
  labels: string[];
  series: ChartSeries[];
  show_legend: boolean;
  show_title: boolean;
  title?: string | null;
}

export type SlideNode = TextNode | ShapeNode | ImageNode | TableNode | ChartNode;

export interface SlideSpec {
  id: string;
  layout?: string | null;
  background: string;
  nodes: SlideNode[];
  notes?: string | null;
}

export interface SlideDeckSpec {
  version: "1.0";
  title: string;
  width: number;
  height: number;
  author: string;
  subject?: string | null;
  warnings: string[];
  slides: SlideSpec[];
}

export function assertSlideDeckSpec(value: unknown): asserts value is SlideDeckSpec {
  if (!value || typeof value !== "object") throw new Error("invalid spec");
  const spec = value as Partial<SlideDeckSpec>;
  exactKeys(spec, ["version", "title", "width", "height", "author", "subject", "warnings", "slides"]);
  if (spec.version !== "1.0" || typeof spec.title !== "string" || !spec.title.trim()) {
    throw new Error("invalid deck metadata");
  }
  if (!positive(spec.width) || spec.width > 100 || !positive(spec.height) || spec.height > 100 ||
      typeof spec.author !== "string" || !Array.isArray(spec.warnings) ||
      !spec.warnings.every((warning) => typeof warning === "string") ||
      !Array.isArray(spec.slides) || !spec.slides.length) {
    throw new Error("invalid deck dimensions or slides");
  }
  for (const slide of spec.slides) {
    exactKeys(slide, ["id", "layout", "background", "nodes", "notes"]);
    if (!slide || typeof slide.id !== "string" || !slide.id.trim() ||
        typeof slide.background !== "string" || !color(slide.background) ||
        !Array.isArray(slide.nodes)) throw new Error("invalid slide");
    for (const node of slide.nodes) {
      if (!node || !["text", "shape", "image", "raster", "table", "chart"].includes(node.type)) {
        throw new Error("invalid node type");
      }
      if (![node.x, node.y].every(nonNegative) || !positive(node.w) || !positive(node.h)) {
        throw new Error("invalid node bounds");
      }
      if (node.x + node.w > spec.width! + 1e-6 || node.y + node.h > spec.height! + 1e-6) {
        throw new Error("node outside slide");
      }
      assertNode(node);
    }
  }
}

function assertNode(node: SlideNode): void {
  const bounds = ["type", "x", "y", "w", "h"];
  switch (node.type) {
    case "text":
      exactKeys(node, [...bounds, "text", "font_size", "font_face", "color", "bold",
        "align", "valign", "margin", "bullet"]);
      if (typeof node.text !== "string" || !node.text.trim() || !positive(node.font_size) ||
          node.font_size > 200 || typeof node.font_face !== "string" || !node.font_face ||
          !color(node.color) || typeof node.bold !== "boolean" ||
          !["left", "center", "right"].includes(node.align) ||
          !["top", "mid", "bottom"].includes(node.valign) ||
          !nonNegative(node.margin) || typeof node.bullet !== "boolean") throw new Error("invalid text node");
      return;
    case "shape":
      exactKeys(node, [...bounds, "shape", "fill", "line_color", "line_width"]);
      if (!["rect", "roundRect", "line", "ellipse"].includes(node.shape) ||
          !color(node.fill) || !color(node.line_color) || !nonNegative(node.line_width)) {
        throw new Error("invalid shape node");
      }
      return;
    case "image":
    case "raster":
      exactKeys(node, [...bounds, "path", "transparency"]);
      if (typeof node.path !== "string" || !node.path.trim() ||
          !nonNegative(node.transparency) || node.transparency > 100) throw new Error("invalid image node");
      return;
    case "table": {
      exactKeys(node, [...bounds, "rows", "font_size", "color", "header_fill",
        "header_color", "border_color"]);
      if (!Array.isArray(node.rows) || !node.rows.length || !Array.isArray(node.rows[0]) ||
          !node.rows[0].length || !node.rows.every((row) =>
            Array.isArray(row) && row.length === node.rows[0].length &&
            row.every((cell) => typeof cell === "string")) ||
          !positive(node.font_size) || node.font_size > 100 ||
          ![node.color, node.header_fill, node.header_color, node.border_color].every(color)) {
        throw new Error("invalid table node");
      }
      return;
    }
    case "chart":
      exactKeys(node, [...bounds, "chart_type", "labels", "series", "show_legend",
        "show_title", "title"]);
      if (!["bar", "column", "line", "pie", "doughnut"].includes(node.chart_type) ||
          !Array.isArray(node.labels) || !node.labels.length ||
          !node.labels.every((label) => typeof label === "string") ||
          !Array.isArray(node.series) || !node.series.length ||
          !node.series.every((series) => typeof series?.name === "string" &&
            Array.isArray(series.values) && series.values.length === node.labels.length &&
            series.values.every((item) => typeof item === "number" && Number.isFinite(item))) ||
          typeof node.show_legend !== "boolean" || typeof node.show_title !== "boolean") {
        throw new Error("invalid chart node");
      }
  }
}

const positive = (value: unknown): value is number =>
  typeof value === "number" && Number.isFinite(value) && value > 0;
const nonNegative = (value: unknown): value is number =>
  typeof value === "number" && Number.isFinite(value) && value >= 0;
const color = (value: unknown): value is string =>
  typeof value === "string" && /^#?[0-9A-F]{6}$/i.test(value);
const exactKeys = (value: object, allowed: string[]): void => {
  if (Object.keys(value).some((key) => !allowed.includes(key))) throw new Error("unexpected field");
};
